pkg/app/carbonapi/http_handlers.go

```diff
@@ -939,39 +940,148 @@ func (app *App) findHandler(w http.ResponseWriter, r *http.Request, lg *zap.Logg
+	}
+	groups[globs.Name] = names
+	data := map[string]interface{}{
+		"results": groups,
+	}
+	if groupByExpr != "1" {
+		flatData := make([]string, 0)
+		for _, group := range groups {
+			flatData = append(flatData, group...)
 		}
+		data["results"] = flatData
+	}
+	err := json.NewEncoder(&b).Encode(data)
+	return b.Bytes(), err
+}
+
+func (app *App) expandHandler(w http.ResponseWriter, r *http.Request, lg *zap.Logger) {
+	t0 := time.Now()
+	defer func() {
+		d := time.Since(t0).Seconds()
+		app.ms.DurationTotal.WithLabelValues("expand").Observe(d)
+		app.ms.ExpandDurationExp.Observe(d)
+		app.ms.ExpandDurationLin.Observe(d)
+	}()
+
+	ctx, cancel := context.WithTimeout(r.Context(), app.config.Timeouts.Global)
+	defer cancel()
+	uuid := util.GetUUID(ctx)
+
+	app.ms.Requests.Inc()
+
+	query := r.FormValue("query")
+	groupByExpr := r.FormValue("groupByExpr")
+	leavesOnly := r.FormValue("leavesOnly")
+	jsonp := r.FormValue("jsonp")
+	useCache := !parser.TruthyBool(r.FormValue("noCache"))
+
+	toLog := carbonapipb.NewAccessLogDetails(r, "expand", &app.config)
+	toLog.Targets = []string{query}
+
+	lg = lg.With(zap.String("request_id", uuid), zap.String("request_type", "expand"), zap.String("target", query))
+	Trace(lg, "received request")
+
+	logLevel := zap.InfoLevel
+	defer func() {
+		if toLog.HttpCode/100 == 2 {
+			if toLog.TotalMetricCount < int64(app.config.ResolveGlobs) {
+				app.ms.ExpandDurationLinSimple.Observe(time.Since(t0).Seconds())
+			} else {
+				app.ms.ExpandDurationLinComplex.Observe(time.Since(t0).Seconds())
+			}
+		}
+		app.deferredAccessLogging(lg, r, &toLog, t0, logLevel)
+	}()
+
+	if query == "" {
+		writeError(uuid, r, w, http.StatusBadRequest, "missing parameter `query`", "", &toLog)
+		return
+	}
+	metrics, fromCache, err := app.resolveGlobs(ctx, query, useCache, &toLog, lg)
+	toLog.FromCache = fromCache
+	if err == nil {
+		toLog.TotalMetricCount = int64(len(metrics.Matches))
 	} else {
-		w.Header().Set("Content-Type", contentType)
-		if _, writeErr := w.Write(blob); writeErr != nil {
-			toLog.HttpCode = 499
-			logLevel = zapcore.WarnLevel
+		var notFound dataTypes.ErrNotFound
+
+		switch {
+		case errors.As(err, &notFound):
+			// we can generate 404 for expand, it's graphite-web-1.0 only
+			Trace(lg, "not found")
+			app.ms.ExpandNotFound.Inc()
+			writeError(uuid, r, w, http.StatusNotFound, err.Error(), "", &toLog)
+			logLevel = zapcore.ErrorLevel
+			return
+		case errors.Is(err, context.DeadlineExceeded):
+			writeError(uuid, r, w, http.StatusUnprocessableEntity, "context deadline exceeded", "", &toLog)
+			logLevel = zapcore.ErrorLevel
+			return
+		default:
+			writeError(uuid, r, w, http.StatusUnprocessableEntity, err.Error(), "", &toLog)
+			logLevel = zapcore.ErrorLevel
 			return
 		}
 	}
 
-	toLog.HttpCode = http.StatusOK
+	if ctx.Err() != nil {
+		app.ms.RequestCancel.WithLabelValues("expand", ctx.Err().Error()).Inc()
+	}
+
+	blob, err := expandEncoder(metrics, leavesOnly, groupByExpr)
```

Comment:
As you can see above - almost everything in expandhandler is just a copy of findHandler, but with different parameters, logging and metrics. Maybe we can simplify this, not sure if we should do that in the same PR, though.

Reply:
Personally, I'd prefer to refactor findHandler first, so that it's easier and less verbose to implement expand. But you've already done it, and if we all agree that the refactoring here is crucial and we will do it afterwards, I think we can move forwards without refactoring at this moment.

Reply:
I agree that the refactor needs to be done but maybe in separate task

Reply:
agree w/ Emad, better to avoid duplication, so makes sense to refactor in same PR or as prev PR

but we can also merge as is and refactor later 🤷

Reply:
I'm not even sure that refactor is suitable here TBH. Yes, handlers look similar but I'm not sure where we can have comfortable level of duplication vs reimplementation of features.
Probably we should keep handlers for find and expand separate, but then we should have separate metrics and logging. What common left to refactor then?